### ISSUE TITLE
Plan to support setup.py and remaining Python dependency lockfiles

### ISSUE THREAD

**Issue by rtfpessoa** [google/deps.dev#102]
Currently the deps.dev UI and APIs have a big percentage of python versions maybe 20-30% which have no known dependencies.
For Python specifically, this is mostly due to lack of support for setup.py, setup.cfg and some other dependency lockfiles.

Is there a plan to support these files?

**Comment by rtfpessoa** [comment]
@cuixq maybe you can help?

**Comment by cuixq** [comment]
@PFCM knows more about this :)

**Comment by rtfpessoa** [comment]
Maybe while I wait for the answer I can provide some details from my side.
We are very interested in this support and are currently adding parsing capabilities to osv-scanner to:
* Support for parsing the `install_requires` key in setup.py and setup.cfg
* In setup.py it only supports arrays of plain string values being passed as a named argument
* Any dependencies described in other requires keys are not scanned
* Fails fast on unsupported inputs

What is the deps.dev source for deps? Does it download the binaries/releases and parses the source with osv-scanner?
If yes, @cuixq do you think osv-scanner would accept a PR to add support for this?

**Comment by cuixq** [comment]
I think deps.dev fetches data from upstream (see https://docs.deps.dev/api/v3/#data) and computes dependency graph based on that.

Contribution is always welcome, so feel free to send the PR to OSV-Scanner!

**Comment by rtfpessoa** [comment]
@cuixq Any idea about the code that does this for python? I tried to look for it but could not find it.

The information, at least in good part only exists in the `sdist` binaries on pypi, not through APIs for this cases.

**Comment by cuixq** [comment]
I don't think the code for Python is open sourced.

**Comment by PFCM** [comment]
Hello, sorry for the delay and thanks for getting in touch.

For reference, just to make sure we're all on the same page: for Python, deps.dev operates on release artifacts downloaded from PyPI, and the dependencies we provide are intended to be as close as practical to the dependencies that pip would install if you were to run `pip install <package>==<version>` on a reasonably modern x86_64 linux system. As Xueqin mentioned the code for that is not open source, although it could be without too much trouble if there's interest.

As you've noticed the biggest cause of missing dependencies is because they're specified in setup.py, this covers about 22% of all the artifacts we've checked (although only 17% of all versions, because many versions have a wheel that is easier to handle). 

Despite this we don't currently have any plans to add support for setup.py specifically. The biggest reasons are:
- this is the first time anyone has asked about it :)
- over the past year, only 3.5% of published releases have used it as the only way to specify dependencies (probably why no one has asked, we don't seem to get too many people interested in historical data for whatever reason)
- relatedly, PEP 517 is starting to take over, which is slightly easier for us (could still ultimately require executing unknown Python code to cover all possible cases, but also PEP 517 builders tend to produce PKG-INFO files that actually have the dependencies in them, so we never need to look at the pyproject.toml etc.)
- it's quite awkward to do well; given the above the value of a partial solution (all that's really feasible without a great deal of effort) seemed low.

However, plans can change. With all that in mind I would be interested to know what the use case is for these more historic dependencies (and/or whether there's an appetite for publishing more of our python code here where we might accept PRs)

**Comment by leifdreizler** [comment]
@PFCM - I _think_ the answer is yes based on:

> the code for that is not open source, although it could be without too much trouble if there's interest.

and

> whether there's an appetite for publishing more of our python code here where we might accept PRs

but if you opensourced this part of deps.dev, does that mean someone from the community could add support/partial support for `setup.py`?

**Comment by leifdreizler** [comment]
Appreciate you replying and tagging folks even though you're no longer at Google 😄 

There's a banner at the top of packages that mentions the lack of deps.dev support, [example](https://deps.dev/pypi/elasticmock)

<img width="728" height="192" alt="Image" src="https://github.com/user-attachments/assets/52694308-0d26-4744-ae92-f378f7191ba8" />

**Comment by PFCM** [comment]
Yeah, if you have a look inside the [tarball](https://pypi.org/project/ElasticMock/1.8.1/#files) the PKG-INFO file has a `License: UNKNOWN` line (although there is a LICENSE file in there too, but what's the right thing to do when those don't match?). There's also no `Requires-Dist`, but the setup.py does have an `install_requires`, which is likely to be where that message is coming from.

That's about all I've got to offer though, it's probably time to recuse myself from this conversation, good luck :)

**Comment by tymzd** [comment]
> I suspect the setup.py support is unlikely to be the problem there, if the license metadata was in the setup.py in an expected place, it would be in the metadata files produced from the setup.py during the build process.

This aligns with my understanding too, although I found that for package versions where [setup.py](http://setup.py/) holds dependencies, no metadata is served at all, including licenses. I’ve rolled out a change to serve partial metadata results in these cases so the packages pointed out by @k80kent should now have license information available.
